### PR TITLE
[TRTLLM-14616][feat] add VisualGen fp8 row-wise quant + autotuner

### PR DESCRIPTION
**Pull request by o-stoner**
## Dev Engineer Review

- Added FP8 per-channel-per-token quantization for VisualGen.
- Added row-wise FP8 weight quantization and loading.
- Added dynamic per-token activation quantization for 3D inputs.
- Added TRT-LLM and vectorized kernel selection with autotuning.
- Added `TRTLLM_FP8_QUANT_TACTIC` overrides.
- Added validation for CUDA inputs, supported dtypes, tensor dimensions, contiguity, and hidden-size limits.
- Added configuration and package exports for `FP8_PER_CHANNEL_PER_TOKEN`.
- Added vendored CUDA vectorization and FP8 compatibility helpers.
- Added explicit kernel instantiations for BF16, FP16, and FP32.
- Forced dynamic activation quantization for the new algorithm.
- No test-list files were included in the provided changes.
- Review should confirm coding-guideline compliance, CUDA compatibility, kernel performance, operator registration, and environment-variable behavior.

## QA Engineer Review

Test-code changes were added.

- Added unit tests for vectorized and tunable per-token FP8 quantization.
- Added tests for output shapes, dtypes, scales, 3D inputs, numerical agreement, tactic overrides, and default tactic selection.
- Added row-wise quantization tests for multiple shapes, round-trip accuracy, and zero-valued weights.
- Added model-loader and configuration parser coverage.
- Extended Cosmos, FLUX, LTX2, Qwen Image, and Wan pipeline coverage to `FP8_PER_CHANNEL_PER_TOKEN`.
- Added end-to-end FP8 accuracy coverage for the Wan pipeline.
- No corresponding `test-db/` or `qa/` registrations were shown for the modified test functions.

Verdict: needs follow-up.

## Description

- Wires FP8_PER_CHANNEL_PER_TOKEN (rowwise FP8) into VisualGen's quantization plumbing. This was already supported by TRT-LLM's FP8RowwiseLinearMethod but was not reachable from the VisualGen pipeline.
- Adds quantize_fp8_rowwise to visual_gen/quantization/ops.py: per-output-row weight quantization (bf16 → fp8 e4m3, one scale per output row) applied at model load time
- Routes FP8_PER_CHANNEL_PER_TOKEN checkpoints through the new weight quantizer in DynamicLinearWeightLoader; pre-quantized fp8 weights pass through unchanged
- Registers FP8_PER_CHANNEL_PER_TOKEN as a valid quant algo string in DiffusionPipelineConfig
- Fixes FP8RowwiseLinearMethod.apply to handle 3D activation inputs (batch, seq, hidden); the rowwise GEMM kernel requires 2D; inputs are flattened before the GEMM and unflattened after
- Adds a vectorized FP8 per-token activation quantization kernel (vendored from vLLM v0.25.0) and an AutoTuner op (`tunable_fp8_per_token_quant`) that profiles the TRT-LLM kernel (`dynamicQuantizeMatrixPerToken`) against it at warmup and caches the winner per input shape; VisualGen enables this via `FP8RowwiseLinearMethod.use_tunable_quantize`; LLM paths leave it disabled; `TRTLLM_FP8_QUANT_TACTIC=trtllm|vectorized` forces a specific kernel

Perf/quality report example for dynamic FP8 row-wise quant on FLUX.2-dev + Wan 2.2 T2V: https://sc.talos.nvidia.com/view/home/scratch.ostoner_coreai_1/TensorRT-LLM-3/TensorRT-LLM/examples/visual_gen/fp8_rowwise_quant_study.html 

## Test Coverage

**New unit tests:**
- tests/unittest/_torch/visual_gen/test_quant_ops.py: quantize_fp8_rowwise op: output dtype, scale shape (out_features,), round-trip error
- tests/unittest/_torch/visual_gen/test_model_loader.py: config parse ("FP8_PER_CHANNEL_PER_TOKEN" → QuantAlgo, dynamic_weight_quant=True); weight loading on Wan 1.3B with weight_scale.dim() == 1 assertion
- tests/unittest/_torch/thop/parallel/test_fp8_per_token_quant_autotuner.py: vectorized kernel: output shape/dtype for bf16/fp16/fp32 inputs, 3D input, scales positive, scale agreement with TRT-LLM kernel; autotuner op: shape/dtype, both tactics numerically equivalent (dequantized), scale dtype normalization to float32, env var forcing (trtllm/vectorized), default tactics include both

**Extended existing tests (parametrize additions):**
- tests/unittest/_torch/visual_gen/test_flux_pipeline.py: FP8_PER_CHANNEL_PER_TOKEN added to: FLUX.1/FLUX.2 weight loading, single-layer numerical correctness vs BF16, full transformer e2e cosine similarity vs BF16
- tests/unittest/_torch/visual_gen/test_ltx2_pipeline.py: weight loading, single-layer correctness, two-stage pipeline loading
- tests/unittest/_torch/visual_gen/test_cosmos3_transformer.py: transformer weight loading
- tests/unittest/_torch/visual_gen/test_qwen_image_pipeline_config.py: config parsing

**Pre-existing tests covering linear.py change:**
- tests/unittest/_torch/thop/parallel/test_fp8_rowwise_linear.py: FP8RowwiseLinearMethod forward correctness (covers the reshape code path indirectly via the full forward call)
- tests/unittest/_torch/visual_gen/test_wan21_t2v_pipeline.py: added wan21_t2v_fp8_rowwise fixture and parametrized it into test_fp8_e2e_accuracy (cos_sim > 0.99 vs BF16)
- tests/unittest/_torch/visual_gen/test_wan22_t2v_pipeline.py: added wan22_t2v_fp8_rowwise fixture covering both transformer stages

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

File: tensorrt_llm/_torch/visual_gen/config.py [hunk 1]
```diff
@@ -601,13 +609,18 @@ def from_pretrained(
...
+        # (full image/video latents) amortize the AutoTuner per-call overhead.
         if quant_config.quant_algo == QuantAlgo.NVFP4:
             from tensorrt_llm._torch.modules.linear import NVFP4LinearMethod
 
             NVFP4LinearMethod.use_tunable_quantize = True
 
+        if quant_config.quant_algo == QuantAlgo.FP8_PER_CHANNEL_PER_TOKEN:
+            from tensorrt_llm._torch.modules.linear import FP8RowwiseLinearMethod
+
+            FP8RowwiseLinearMethod.use_tunable_quantize = True
```
[the remark this answers is not in the document]
Comment (o-stoner) [reply tensorrt_llm/_torch/visual_gen/config.py:641]: Won't fix; there should be no deployment path where this class mutation affects LLM inference

File: tests/unittest/_torch/visual_gen/test_wan22_t2v_pipeline.py [hunk 2]
```diff
@@ -308,6 +308,17 @@ def wan22_t2v_fp8():
...
+def wan22_t2v_fp8_rowwise():
+    pipeline = _make_wan22_t2v(
+        quant_config={"quant_algo": "FP8_PER_CHANNEL_PER_TOKEN", "dynamic": True}
+    )
+    yield pipeline
+    del pipeline
+    gc.collect()
+    torch.cuda.empty_cache()
+
+
```
[the remark this answers is not in the document]
Comment (o-stoner) [reply tests/unittest/_torch/visual_gen/test_wan22_t2v_pipeline.py]: Fixed; removed the wan22_t2v_fp8_rowwise fixture for now

File: tests/unittest/_torch/thop/parallel/test_fp8_per_token_quant_autotuner.py [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+import torch
+
+
+class TestVectorizedPerTokenFP8Quant(unittest.TestCase):
+    """Test the vectorized FP8 per-token quantization kernel."""
+
+    def setUp(self):
+        import tensorrt_llm  # noqa: F401 — registers C++ ops
+
+        torch.manual_seed(42)
```
[the remark this answers is not in the document]
Comment (o-stoner) [reply tests/unittest/_torch/thop/parallel/test_fp8_per_token_quant_autotuner.py:16]: Won't fix. This project's tests require a CUDA environment.

File: tests/unittest/_torch/visual_gen/test_quant_ops.py [hunk 4]
```diff
@@ -115,6 +116,43 @@ def test_fp8_blockwise_zero_weight(self):
...
+                self.assertEqual(qweight.shape, weight.shape)
+                self.assertEqual(scale.shape, (shape[0],))
+
+    def test_fp8_rowwise_zero_weight(self):
+        """Test FP8 rowwise quantization with zero weight."""
+        weight = torch.zeros(128, 256, dtype=torch.bfloat16, device="cuda")
+        qweight, scale = quantize_fp8_rowwise(weight)
+
+        self.assertEqual(qweight.dtype, torch.float8_e4m3fn)
+        self.assertTrue(torch.all(qweight.to(torch.float32) == 0))
```
[the remark this answers is not in the document]
Comment (o-stoner) [reply tests/unittest/_torch/visual_gen/test_quant_ops.py:154]: Won't fix; did not fail lint job

File: tensorrt_llm/_torch/modules/linear.py [hunk 5]
```diff
@@ -981,6 +981,11 @@ class FP8RowwiseLinearMethod(UnquantizedLinearMethod):
 
     supports_nccl_symmetric_memory_window_output: ClassVar[bool] = True
 
+    # When True, use tunable_fp8_per_token_quant (AutoTuner selects TRT-LLM vs
+    # vectorized kernel). Visual gen pipelines set this to True before model
+    # construction; LLM paths leave it False.
+    use_tunable_quantize: bool = False
```
[the remark this answers is not in the document]
Comment (o-stoner) [reply tensorrt_llm/_torch/modules/linear.py:1032]: Follows the existing NVFP4LinearMethod pattern; will not fix

File: cpp/tensorrt_llm/thop/fp8Op.cpp [hunk 6]
```diff
@@ -372,6 +373,49 @@ Tensor symmetric_dequantize_per_tensor(Tensor input, Tensor scales)
...
+    TORCH_CHECK(_st == torch::kFloat32 || _st == torch::kFloat16 || _st == torch::kBFloat16,
+        "vectorized_per_token_fp8_quant: input must be FP16, BF16, or FP32");
+
+    int64_t const numel = input.numel();
+    int64_t const hidden_size = input.size(-1);
+    int64_t const num_tokens = numel / hidden_size;
+    TORCH_CHECK(hidden_size <= INT_MAX, "hidden_size exceeds INT_MAX");
+
+    Tensor output
+        = torch::empty(input.sizes(), torch::dtype(torch::kFloat8_e4m3fn).device(torch::kCUDA).requires_grad(false));
```
Comment (chang-l) [inline cpp/tensorrt_llm/thop/fp8Op.cpp:397]: These allocations use the current CUDA device, while the launch on the input GPU `input.get_device()`. Could we have a CUDA guard for this?

File: tensorrt_llm/_torch/custom_ops/torch_custom_ops.py [hunk 7]
```diff
@@ -2908,3 +2908,82 @@ def _(
...
+    _, best_tactic = tuner.choose_one(
+        "trtllm::fp8_per_token_quant_tactic",
+        [runner],
+        Fp8PerTokenQuantRunner.tuning_config,
+        [x],
+    )
+    qx, scale = runner(inputs=[x], tactic=best_tactic)
+    # Normalize scale to float32 regardless of tactic: TRTLLM tactic returns
+    # input-dtype scales (e.g. bf16); vectorized tactic already returns float32.
+    return [qx, scale.float()]
```
Comment (chang-l) [inline tensorrt_llm/_torch/custom_ops/torch_custom_ops.py]: Could we include the scale normalization in the profiled runner? For BF16 input, TRTLLM returns a BF16 scale and pays `scale.float()` after `choose_one`, while VECTORIZED already returns FP32. The tuner therefore omits a real, tactic-specific cast from only the TRTLLM timing and can cache the wrong winner. Returning `scale.float()` from `Fp8PerTokenQuantRunner.forward()` would time the same work inference executes.

File: cpp/tensorrt_llm/kernels/fp8PerTokenQuant/fp8_per_token_quant.cu [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+TRTLLM_NAMESPACE_BEGIN
+
+namespace kernels
+{
+
+template <typename T_IN>
+void invokeVectorizedPerTokenFP8Quant(
+    void* output, float* scales, T_IN const* input, int hidden_size, int64_t num_tokens, cudaStream_t stream)
+{
+    dim3 const grid(static_cast<unsigned>(num_tokens));
```
[the remark this answers is not in the document]
Comment (o-stoner) [reply cpp/tensorrt_llm/kernels/fp8PerTokenQuant/fp8_per_token_quant.cu:32]: Resolved;  TORCH_CHECK(num_tokens <= INT_MAX, ...) at fp8Op.cpp:392

File: tensorrt_llm/_torch/custom_ops/torch_custom_ops.py [hunk 9]
```diff
@@ -2908,3 +2908,82 @@ def _(
...
+    """
+    tuner = AutoTuner.get()
+    runner = Fp8PerTokenQuantRunner()
+    _, best_tactic = tuner.choose_one(
+        "trtllm::fp8_per_token_quant_tactic",
+        [runner],
+        Fp8PerTokenQuantRunner.tuning_config,
+        [x],
+    )
+    qx, scale = runner(inputs=[x], tactic=best_tactic)
```
[the remark this answers is not in the document]
Comment (o-stoner) [reply tensorrt_llm/_torch/custom_ops/torch_custom_ops.py]: Fixed: env-var override (TRTLLM_FP8_QUANT_TACTIC) dispatched before choose_one() at torch_custom_ops.py:2965-2980

File: cpp/tensorrt_llm/thop/fp8Op.cpp (hunk 6 above)
Comment (o-stoner) [reply cpp/tensorrt_llm/thop/fp8Op.cpp:397]: Thanks, added `at::cuda::CUDAGuard device_guard{...}` at `fp8Op.cpp:394`

File: tensorrt_llm/_torch/custom_ops/torch_custom_ops.py (hunk 7 above)
Comment (o-stoner) [reply tensorrt_llm/_torch/custom_ops/torch_custom_ops.py]: Yes, thanks. `scale.float()` is now moved inside `Fp8PerTokenQuantRunner.forward()` (`torch_custom_ops.py:2951`)

**Review by mikeiovine — APPROVED** [review]

**Review by mzweilz — APPROVED** [review]
LGTM from the infra side.

**Review by allisonlim-nv — APPROVED** [review]

**Review by rosong11 — APPROVED** [review]
LGTM
